d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -547,6 +573,42 @@ private void onData(ResourceUpdate data)
       }
     }
 
+    // track rough estimate of latency spent on the xds server in millis = resource receipt time - resource modified time
+    private void trackServerLatency(ResourceUpdate resourceUpdate, XdsServerMetricsProvider metricsProvider)
+    {
+      if (_data == null)
+      {
+        return;
+      }
+
+      long now = SystemClock.instance().currentTimeMillis();
+      if (resourceUpdate instanceof NodeUpdate)
+      {
+        XdsD2.Node nodeData = ((NodeUpdate) resourceUpdate).getNodeData();
+        if (nodeData == null)
+        {
+          return;
+        }
+        metricsProvider.trackLatency(now - nodeData.getStat().getMtime());
+      }
+      else if (resourceUpdate instanceof D2URIMapUpdate)
+      {
+        // only track server latency for the updated/new uris in the update
+        Map<String, XdsD2.D2URI> currentUriMap = ((D2URIMapUpdate) _data).getURIMap();
+        MapDifference<String, XdsD2.D2URI> rawDiff = Maps.difference(((D2URIMapUpdate) resourceUpdate).getURIMap(),
+            currentUriMap == null ? Collections.emptyMap() : currentUriMap);
+        List<XdsD2.D2URI> updatedUris = rawDiff.entriesDiffering().values().stream()
+            .map(ValueDifference::leftValue) // new data of updated uris
+            .collect(Collectors.toList());
+        updatedUris.addAll(rawDiff.entriesOnlyOnLeft().values()); // newly added uris
+
+        for (XdsD2.D2URI uri : updatedUris)
+        {
+          metricsProvider.trackLatency(now - uri.getModifiedTime().getSeconds() * 1000);
```

Comment:
You should avoid actually materializing this list. The google stuff only creates views for these diffs, it doesn't actually create the map, so you should iterate over them without creating the list

Reply:
good catch. Fixed, thx.